Link to Events in the Past

daizylogik opened the pull request:
@davidhabib Hi David,
Here are my changes related to displaying events in the past, if linked to directly. Please let me know if you have any questions.
Thanks.
Vladimir

davidhabib commented on src/classes/VOL_CTRL_VolunteersJobListingFS.cls:
```diff
@@ -270,7 +271,7 @@ global virtual with sharing class VOL_CTRL_VolunteersJobListingFS {
                         (Select Id, Start_Date_Time__c, Duration__c, Number_of_Volunteers_Still_Needed__c, Total_Volunteers__c,
                             Description__c, System_Note__c From Volunteer_Job_Slots__r 
-                            where Start_Date_Time__c >= :dtNow and Start_Date_Time__c < :dtLast
+                            //where Start_Date_Time__c >= :dtNow and Start_Date_Time__c < :dtLast
```
I'm concerned that by removing this date filter, you could be returning too many records in this soql query for a job that has daily shifts, and has been in place for several years.  Also, it appears that not only are you showing all past shifts, you are also getting all future shifts, rather than respecting the dtLast value that is specified thru URL parameters.

daizylogik replied:
Returning too many records is a valid concern. One suggestion I have is: put the Start_Date_Time__c < :dtLast condition back, so we don't return any future shifts. In addition to to that, we can also ad a limit to the number of past shifts returned by adding  LIMIT 20, for example. Let me know what you think. Thanks.

davidhabib commented on src/classes/VOL_CTRL_VolunteersJobListingFS.cls:
```diff
@@ -301,12 +302,29 @@ global virtual with sharing class VOL_CTRL_VolunteersJobListingFS {
+                	Volunteer_Job__c job = listVolunteerJobs[0];
+                	eventInThePast = 0 < job.Number_of_Shifts__c;
+                	for (Volunteer_Shift__c shift : job.Volunteer_Job_Slots__r) {
+                		if (date.today() <= shift.Start_Date_Time__c) {
```
seems like this logic will say the job is an event in the past, even if it has multiple shifts, some of which are old and some of which are new.  am I misunderstanding this?

davidhabib replied:
correction.  seems like it is saying if any shifts are in the future, then the job as no events in the past, which seems incorrect.

daizylogik replied:
The purpose of this property is to return true only when all all the shifts are in the past, and the job has no shifts today or in the future. I can rename the property to allShiftsInThePast, or noFutureShifts. Please let me know if you prefer another name.

davidhabib commented:
I haven't had time to pull this code down and test it, but I'm not sure it does what I expected.  I thought it would only change the page's behavior such that if a specific shift is selected, it would display it even though it was in the past, and it would not provide the signup link.

did you have another intention in mind?  thanks.

sgjudd commented:
David, One thing we are trying to accomplish is to allow a job to have a permanent link with only the jobId being passed (e.g., http://v4sapp-developer-edition.na3.force.com/VolunteersJobListingFS?jobId=a025000000ce2BK) In this example, you can see there was a prior shift, as well as a future shift which you can sign up for. The job itself is not marked as past, since there is an available upcoming shift.

I agree with your concern about jobs with a large number of shifts returning to many records. Do you have ideas about the best way to handle it? We need to show at least some shift information, so that viewers would know when the job occurred.

daizylogik commented:
Steve and I had talked about this issue. Please let us know if you have any suggestions on how to address this, and if the suggestions I had above regarding limiting the number of records returned are acceptable.

davidhabib commented:
some questions:
1. if the job only had old shifts, do you still want the old shifts shown (with closed), or were you hoping to see no shifts but just display a message under the job that there are no upcoming shifts?  I believe the latter is what Vladmir currently implemented.

2. maybe the better way to approach this is a page parameter that specifies how many months of shifts in the past to display, similar to how we do how many months into the future?  This way the job listing could display a finite amount of past shifts.  Unfortunately, if your shift granularity is once every 3 months, and your nMonthsPastToShow parameter is 1, you would see those past shifts.  I'm hesitant to suggest a nShiftsPastToShow parameter, since that would still require the query to return all past shifts, and just the server logic to then remove the older shifts. thus the query could still run into issues.  On the other hand, if this parameter was only supported if filtering on a single Job, the chance of overrunning query limits would be much less.  But then you know people will want to use it in all scenarios, not just when filtering on a specific job!

thoughts?

sgjudd commented:
David,

I spoke with Vladimir, and we're going to take a slightly different tack, to alleviate any concerns about the SOQL returning too many shifts. I've asked Vladimir to select the "First Shift" field from the Volunteer Job. If a job has upcoming shifts, the job will display as it does now. If there are no upcoming shifts, then the job will still display with the following changes: a CSS clss added to enable a different styling for past events and a message: This job occurred in the past. First Shift: DATE

I'd like a date on the page to provide a bit of context about when the job occurred - was it yesterday or two years ago?

What do you think?

davidhabib commented:
sounds good!

daizylogik commented:
David,
I refactored the code as agreed: the where clause in the soql for the shifts is back. I also renamed the property for past events to eventAllShiftsInThePast to make it more accurate.
Let me know if there's anything else you would like me to change.
Thanks. Vladimir

davidhabib commented on src/labels/CustomLabels.labels:
```diff
@@ -432,6 +432,14 @@ day:      &apos;day&apos;</value>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>labelEventInThePast</shortDescription>
+        <value>This event is not accepting signups at this time.</value>
```
Please change this to "This job has no upcoming shifts."

V4S doesn't use the terminology Event.  I realize UNH wants it to be event, but they can certainly translate the label.

davidhabib commented on src/classes/VOL_CTRL_VolunteersJobListingFS.cls:
```diff
@@ -56,6 +56,7 @@ global virtual with sharing class VOL_CTRL_VolunteersJobListingFS {
     global integer nMonthsToShow { get; set; }
     global integer nDaysToShow { get; set; }
     global date dtMonthFilter { get; set; }
+    global boolean eventAllShiftsInThePast { get; set; }
```
please change this to jobAllShiftsInThePast, since V4S does not use the terminology Event.  thanks.

davidhabib commented on src/pages/VolunteersJobListingFS.page:
```diff
@@ -122,7 +122,9 @@
-    	<div id="jobContainer" class="{!IF(job.External_Signup_Url__c == null, '', 'externalJob')}">
+    	<apex:variable var="containerClass" value="{!IF(job.External_Signup_Url__c != null, 'externalJob', IF(eventAllShiftsInThePast, 'pastEvent', ''))}" />
+    	<div id="jobContainer" class="{!containerClass}">
+    	<!-- <div id="jobContainer" class="{!IF(job.External_Signup_Url__c == null, '', 'externalJob')}">  -->
```
you can remove this commented out line.

davidhabib commented on src/pages/VolunteersJobListingFS.page:
```diff
@@ -156,11 +158,17 @@
+
+        <apex:outputPanel id="panelEventInThePast" rendered="{!eventAllShiftsInThePast}" >
+        	<apex:outputText value="{!$Label.labelEventInThePast}" />
+        	<apex:outputText rendered="{!job.First_Shift__c != null}" value=" First Shift: " />
```
"First Shift:" needs to be a custom label so it can be localized or overridden.  thanks.

davidhabib commented on src/pages/VolunteersJobListingFS.page:
```diff
@@ -156,11 +158,17 @@
+        <apex:outputPanel id="panelEventInThePast" rendered="{!eventAllShiftsInThePast}" >
+        	<apex:outputText value="{!$Label.labelEventInThePast}" />
+        	<apex:outputText rendered="{!job.First_Shift__c != null}" value=" First Shift: " />
+        	<apex:outputText rendered="{!job.First_Shift__c != null}" value="{0,date,MM/dd/yyyy}"><apex:param value="{!job.First_Shift__c}" /></apex:outputText>
```
You need to do the correct date formatting based on the language string that is specified as a URL parameter.  I have a javascript helper on this page to do this.  look for this call in the code to see how to use it:

var strStartDate = MyDateFormat(dtStart, '{!JSENCODE(strDateFormat)}');

daizylogik commented:
David, I made the requested changes, let me know if there's anything else. Thanks. Vladimir
